lxd/profiles_utils.go

```diff
@@ -0,0 +1,135 @@
+package main
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/lxc/lxd/shared/api"
+)
+
+func doProfileUpdate(d *Daemon, name string, id int64, profile *api.Profile, req api.ProfilePut) Response {
+	// Sanity checks
+	err := containerValidConfig(d, req.Config, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	err = containerValidDevices(req.Devices, true, false)
+	if err != nil {
+		return BadRequest(err)
+	}
+
+	containers := getContainersWithProfile(d, name)
+
+	// Check if the root device is supposed to be changed or removed.
+	oldProfileRootDiskDeviceKey, oldProfileRootDiskDevice := containerGetRootDiskDevice(profile.Devices)
+	_, newProfileRootDiskDevice := containerGetRootDiskDevice(req.Devices)
+	if len(containers) > 0 &&
+		oldProfileRootDiskDevice["pool"] != "" &&
+		newProfileRootDiskDevice["pool"] == "" ||
+		(oldProfileRootDiskDevice["pool"] != newProfileRootDiskDevice["pool"]) {
+		// Check for containers using the device
+		for _, container := range containers {
+			// Check if the device is locally overridden
+			localDevices := container.LocalDevices()
+			k, v := containerGetRootDiskDevice(localDevices)
+			if k != "" && v["pool"] != "" {
+				continue
+			}
+
+			// Check what profile the device comes from
+			profiles := container.Profiles()
+			for i := len(profiles) - 1; i >= 0; i-- {
+				_, profile, err := dbProfileGet(d.db, profiles[i])
+				if err != nil {
+					return InternalError(err)
+				}
+
+				// Check if we find a match for the device
+				_, ok := profile.Devices[oldProfileRootDiskDeviceKey]
+				if ok {
+					// Found the profile
+					if profiles[i] == name {
+						// If it's the current profile, then we can't modify that root device
+						return BadRequest(fmt.Errorf("At least one container relies on this profile's root disk device."))
+					} else {
+						// If it's not, then move on to the next container
+						break
+					}
+				}
+			}
+		}
+	}
```

Comment:
Another note to myself to go through this chunk again in the morning. It looks like my suggested pseudo-code but my eyes are tired and I'll need to read through again to be sure that all the old/new logic is right.